projects/ADuCM3029_demo_cn0428_cn0429/src/ADuCM3029_demo_cn0428_cn0429.cpp

```diff
@@ -0,0 +1,1307 @@
+
+/* Memory for the GPIO driver */
+uint8_t gpioMemory[ADI_GPIO_MEMORY_SIZE] = {0};
+
+/* Handle for the UART device. */
+ADI_UART_HANDLE hDevice;
+
+/* Memory for the UART driver. */
+static uint8_t UartDeviceMem[ADI_UART_MEMORY_SIZE];
+
+/* ================================================== */
+/* 						VARIABLES					  */
+/* ================================================== */
+const ADI_GPIO_PORT CS_Port [4] = {ADI_GPIO_PORT0, ADI_GPIO_PORT1, ADI_GPIO_PORT2, ADI_GPIO_PORT0};		// Store Chip Select GPIO Port address
+const ADI_GPIO_DATA CS_Pin [4] = {ADI_GPIO_PIN_8, ADI_GPIO_PIN_11, ADI_GPIO_PIN_1, ADI_GPIO_PIN_9};		// Store Chip Select GPIO Pin address
+const ADI_GPIO_PORT INT_Port [4] = {ADI_GPIO_PORT0, ADI_GPIO_PORT0, ADI_GPIO_PORT1, ADI_GPIO_PORT1};	// Store Interrupt GPIO Port address
+const ADI_GPIO_DATA INT_Pin [4] = {ADI_GPIO_PIN_15, ADI_GPIO_PIN_13, ADI_GPIO_PIN_12, ADI_GPIO_PIN_14};	// Store Interrupt GPIO Pin address
+
+/* Ringbuffer structure */
+struct RingBuf RX, TX;
+
+const uint8_t M355_Demo_Msg[64]		= "-------- M355 Gas Sensing Demo --------";
+const uint8_t M355_Demo_Msg_Wq[64]	=
+	"-------- M355 Water Quality Demo --------";
+
+uint8_t 			TXbuff[256]	= {0};		// UART transmit buffer
+uint8_t 			gBuff[64]	= {0};		// general use buffer
+volatile uint8_t	TXcompleteFlag = 1;		// indicates UART TX complete
+uint8_t				RXchar = 0;				// character received over UART (user input)
+uint8_t				cmdInString[64] = "";	// holds incoming user command
+uint8_t				cmdInCnt = 0;			// counts incoming bytes over UART
+
+M355_GAS			m355_gas_sensor;
+Gas_Reading			*pGasSensor = &m355_gas_sensor;
+
+uint8_t rcal = 0;
+uint8_t pulseResultBuffer [6144] = {0};		// buffer holding pulse test result
+uint8_t EISResponseBuff[1024] = {0};		// buffer holding EIS results
+
+uint8_t *Slave_Rx_Buffer =
+	pulseResultBuffer; //borrow this memory space for water quality results
+uint16_t Slave_Rx_Index = 0;
+uint8_t UART_Temp_Buffer [64] = {0};
+uint8_t UART_Temp_Buffer_Index = 0;
+uint8_t initialcycle = 1;
+
+const char* rtia[] 	= { "0R", "200R", "1k", "2k", "3k", "4k", "6k", "8k", "10k", "12k", "16k", "20k",
+			    "24k", "30k", "32k", "40k", "48k", "64k", "85k", "96k", "100k", "120k",
+			    "128k", "160k", "196k", "256k", "512k"
+		      };
+
+const char* rload[]	= { "0R", "10R", "30R", "50R", "100R", "1k6", "3k1", "3k6"};
+
+bool S1_detected = false;
+bool S2_detected = false;
+bool S3_detected = false;
+bool S4_detected = false;
+
+uint16_t ledTick = 0;
+
+uint16_t streamTickCnt =
+	0;		// used for timing the data update rate when in STREAM state
+
+uint16_t streamTickCfg =
+	1;		// used as configuration parameter for data update rate in STREAM mode
+// default value is 1, i.e. 1 second update rate
+
+eFSM_State FSM_State = INIT;
+
+/* ================================================== */
+/* 						FUNCTIONS					  */
+/* ================================================== */
+
+/* General Delay */
+void delay_ms (uint32_t mSec)
```

Comment:
delay functions can be implemented more precisely  using the Systick and a timer. 
https://github.com/analogdevicesinc/EVAL-ADICUP3029/blob/master/projects/ADuCM3029_demo_cn0326/src/Timer.c
This example will work for delay_ms but i think you need a timer for delay_us.